core/nodefactory.lua

```diff
@@ -276,6 +275,11 @@ function nodefactory.discretionary:toText ()
 end
 
 function nodefactory.discretionary:outputYourself (typesetter, line)
+  -- TODO this is an indication of a deeper bug. If we were asked to output
+  -- discretionaries but the parent nnode is not hyphenated then we're
+  -- outputting things that were only used for measuring "what if" scenarios in
+  -- break point calculations. These nodes shouldn't exist at this point.
+  if self.parent and not self.parent.hyphenated then return end
```

Comment:
Re: "These nodes shouldn't exist at this point": There's nothing (currently) to remove unused discretionaries. The last one in a line is marked as used (the next test below), in `computeLineRatio` which also sets the parent's hyphenated to true in that case.

Reply:
I wasn't expected them to be *removed* necesesarily, but they are getting added at some points they should not be. For example somehow after typesetting a line we're ending up with hyphenation discretionaries *after* each word. A word with 3 hyphenation points that isn't even a candidate for hyphenation will end up with 3 discretionaries *after* the word (not in their places). The actual words near the end of a line that are hyphenation candidates based on having filled up a line and gone over-length will have the correct placement.

Reply:
I split out my findings on this issue in #1528.